data/env/snapd.sh.in

```diff
@@ -1,14 +1,17 @@
 #!/bin/sh --this-shebang-is-just-here-to-inform-shellcheck--
 
 # Expand $PATH to include the directory where snappy applications go.
-if [ "${PATH#*/snap/bin}" = "${PATH}" ]; then
+if [ -n "${PATH##*@SNAP_MOUNT_DIR@/bin*}" ]; then
     export PATH=$PATH:@SNAP_MOUNT_DIR@/bin
 fi
 
 # desktop files (used by desktop environments within both X11 and Wayland) are
 # looked for in XDG_DATA_DIRS; make sure it includes the relevant directory for
 # snappy applications' desktop files.
-if [ "${XDG_DATA_DIRS#*/snapd/desktop}" = "${XDG_DATA_DIRS}" ]; then
-    export XDG_DATA_DIRS="${XDG_DATA_DIRS:-/usr/local/share:/usr/share}:/var/lib/snapd/desktop"
+if [ -z "$XDG_DATA_DIRS" ]; then
+    export XDG_DATA_DIRS="/usr/local/share:/usr/share"
+fi
+if [ -n "${XDG_DATA_DIRS##*/var/lib/snapd/desktop*}" ]; then
```

Comment:
Is the double `#` expected? It does not seem to work in practice.

Reply:
```$ X=FooBar; echo ${X##Foo}
Bar```

works in bash and dash.

Reply:
Note: this syntax doesn't work for empty variables, but it's guarded thanks to the -z check and filing up before in both cases.

FYI, I pushed some similar fixes both in xorg and gnome-session in bionic. (and yeah, it's not a bashism)

Reply:
Same question here, and also the note above should be added to the code. It's too easy to miss the details here and the next person might take the -z off without realizing.

Reply:
Same answer, indeed. In practice, our Xorg and wayland scripts following that logic don't encounter cases where this happens (and it avoids crashing electron for instance, due to a string too long after many logout/logins). I'm happy if we go the subprocess `egrep` way either.

On the note, what about "Ensuring default system xdg paths are set"?